oc_mirror.sh

```diff
@@ -0,0 +1,71 @@
+#!/usr/bin/env bash
+set -ex
+
+# Utility functions to create a local mirror registry using 'mirror-registry' and mirror
+# a release using 'oc mirror'
+
+function add_auth_to_pull_secret() {
+ 
+   quay_auths=`echo -n "${QUAY_REGISTRY_USER}:${QUAY_REGISTRY_PASS}" | base64 -w0`
+
+   tmpauthfile=$(mktemp --tmpdir "quayauth--XXXXXXXXXX")
+   _tmpfiles="$_tmpfiles $tmpauthfile"
+   tmppullsecret=$(mktemp --tmpdir "pullsecret--XXXXXXXXXX")
+   _tmpfiles="$_tmpfiles $tmppullsecret"
+
+
+   cat > "${tmpauthfile}" << EOF
+{
+  "auths": {
+    "${LOCAL_REGISTRY_DNS_NAME}:${LOCAL_REGISTRY_PORT}": {
+      "auth": "$quay_auths"
+    }
+  }
+}
+EOF
+
+   jq -s '.[0] * .[1]' ${tmpauthfile} ${PULL_SECRET_FILE} > ${tmppullsecret}
+   cp ${tmppullsecret} ${PULL_SECRET_FILE}
+
+   # Note that this must also be manually added to .docker/config.json, we don't currently do this in dev-scripts to avoid writing
+   # to user directory.
+
+}
+
+function setup_quay_mirror_registry() {
+
+   mkdir -p ${WORKING_DIR}/oc-mirror
+   pushd ${WORKING_DIR}/mirror-registry
+   sudo ./mirror-registry install --quayHostname virthost.ostest.test.metalkube.org --quayRoot ${WORKING_DIR}/oc-mirror/ --initUser ${QUAY_REGISTRY_USER} --initPassword ${QUAY_REGISTRY_PASS} --sslCheckSkip -v
+
+   add_auth_to_pull_secret
+   popd
+}
+
+function publish_mirror_registry() {
+
+   # Create imageset containing the local URL and the OCP release to mirror
+   tmpimageset=$(mktemp --tmpdir "imageset--XXXXXXXXXX")
+   _tmpfiles="$_tmpfiles $tmpimageset"
+
+   cat > "${tmpimageset}" << EOF
+apiVersion: mirror.openshift.io/v1alpha1
+kind: ImageSetConfiguration
+archiveSize: 4
+storageConfig:
+  registry:
+    imageURL: ${LOCAL_REGISTRY_DNS_NAME}:${LOCAL_REGISTRY_PORT}/origin:latest
+    skipTLS: true
+mirror:
+  ocp:
+    channels:
+    - name: candidate-${OPENSHIFT_RELEASE_STREAM}
+  additionalImages:
+  - name: registry.redhat.io/ubi8/ubi:latest
+EOF
+
+   # TODO - print message on failure that password for must be added to .docker/config.json
```

Comment:
I think we should improve the user experience here. When a user enables OC_MIRROR=true and sees an UNAUTHORIZED.. error, they will not have any idea of how to fix it unless they happen to come across this part of the code. 

I think additional documentation should be added to config_example.sh for OC_MIRROR to explain what needs to be added to .docker/config.json. Since the password is hard-coded in agent/common.sh, config_example.sh should explain that this should be added to .docker/config.json:

`"virthost.ostest.test.metalkube.org:8443": { "auth": "c3RhY2s6cXVheXBhc3N3b3Jk" },`

I think the secrets for registry.redhat.io and quay.io also need to be added to .docker/config.json, and this should be explained in config_example.sh for OC_MIRROR. When they weren't present, I was seeing these errors:

```
+(/home/rwsu/go/src/github.com/openshift-metal3/dev-scripts/oc_mirror.sh:59): publish_mirror_registry(): cat
+(/home/rwsu/go/src/github.com/openshift-metal3/dev-scripts/oc_mirror.sh:78): publish_mirror_registry(): oc mirror --dest-skip-tls --config /tmp/imageset--Blj8LVZGWj docker://virthost.ostest.test.metalkube.org:8443
Checking push permissions for virthost.ostest.test.metalkube.org:8443
Creating directory: oc-mirror-workspace/src/publish
Creating directory: oc-mirror-workspace/src/v2
Creating directory: oc-mirror-workspace/src/charts
Creating directory: oc-mirror-workspace/src/release-signatures
No metadata detected, creating new workspace
level=info msg=trying next host error=failed to authorize: failed to fetch anonymous token: unexpected status: 401 Unauthorized host=registry.redhat.io
error: failed to authorize: failed to fetch anonymous token: unexpected status: 401 Unauthorized
```

`error: unable to retrieve source image quay.io/openshift-release-dev/ocp-v4.0-art-dev manifest sha256:6b357378e32d5f07537b6a657f53b332ac89c7e3800ca4281d59afadaf7cbff3: unauthorized: access to the requested resource is not authorized
error: unable to retrieve source image quay.io/openshift-release-dev/ocp-v4.0-art-dev manifest sha256:9a6435cb4f9394d7f10ae6d25478093e0bb58ee6188c942808d8ecc91c6118a3: unauthorized: access to the requested resource is not authorized
`

Reply:
Good points. I've updated config_example.sh to document how to update .docker/config.json.

As we talked about, there is an RFE to pass the auth into the oc-mirror command https://issues.redhat.com/browse/OCPBUGS-241. If that is implemented we can remove this note as we won't need to updated config.json